sc-memory/sc-memory/sc_agent.hpp

```diff
@@ -0,0 +1,610 @@
+          <= 1;
+    };
+
+    // Each agent should have no more than one of three methods working with result condition.
+    struct ResultConditionMethod
+    {
+      static constexpr bool value =
+          (HasOverride<TScAgent>::CheckResultCondition::value + HasOverride<TScAgent>::GetInitiationCondition::value
+           + HasOverride<TScAgent>::GetInitiationConditionTemplate::value)
+          <= 1;
+    };
+  };
+
+  template <typename TScAgent>
+  struct HasOneOverride
+  {
+  public:
+    struct DoProgramMethod
+    {
+      static bool const value = HasOverride<TScAgent>::DoProgramWithEventArgument::value
+                                || HasOverride<TScAgent>::DoProgramWithoutEventArgument::value;
+    };
+  };
+
+  friend class ScModule;
+  friend class ScActionInitiatedAgent;
+  friend class ScAgentContext;
+
+private:
+  /*!
+   * @brief Subscribes agent class to specified sc-events.
+   *
+   * If provided agent implementation `agentImplementationAddr` is valid then this method searches in knowledge base the
+   * following constructions
+   *
+   * @code
+   *             nrel_inclusion
+   *                  |
+   *                  |
+   *                  |
+   *                  \/
+   * abstract_agent ======> agent_implementation
+   *
+   *   nrel_primary_initiation_condition
+   *                  |
+   *                  |
+   *                  |  event_class
+   *                  \/     ||
+   * abstract_agent =======> ||
+   *                         ||
+   *                         \/
+   *               event_subscription_element
+   * @endcode
+   *
+   * and subscribes the agent of this class to sc-event `event_class` with subscription sc-element
+   * `event_subscription_element`, else subscribes the agent of this class to sc-event
+   * `TScEvent` with subscription sc-elements from `subscriptionAddrs`.
+   *
+   * @tparam TScAgent An agent class to be subscribed from the event.
+   * @param context A sc-memory context used to subscribe agent class to specified sc-event.
+   * @param agentImplementationAddr A sc-address of agent implementation specified in knowledge base for this agent
+   * class.
+   * @param subscriptionAddrs A list of sc-addresses of sc-elements to subscribe to.
+   * @warning Specified agent class must be derived from class `ScAgent`.
+   * @throws utils::ExceptionInvalidParams if any of the subscription addresses are invalid.
+   * @throws utils::ExceptionInvalidState if the agent is already subscribed to the event.
+   * @throws utils::ExceptionInvalidState if the agent implementation for this agent class is valid and is not included
+   * in any abstract sc-agent.
+   * @throws utils::ExceptionInvalidState if the agent implementation for this agent class is valid and the abstract
+   * sc-agent for this agent class does not have a primary initiation condition.
+   */
+  template <class TScAgent, class... TScAddr>
+  static _SC_EXTERN void Subscribe(
```

Comment:
on my pov methods Subscribe, Unsubscribe, GetCallback are methods for another entity like ScAgentManager. This logic is not about the work of agents, but about managing their subscriptions. And if you take this logic out of here, then there will be nothing left in ScAgent and you will not have to think about the difference between ScAgentBase and ScAgent

Reply:
Fixed